core/src/main/scala/kafka/server/KafkaConfig.scala

```diff
@@ -1892,8 +1892,9 @@ class KafkaConfig(val props: java.util.Map[_, _], doLog: Boolean, dynamicConfigO
     }
   }
 
+  // Topic IDs are used with all self-managed quorum clusters and ZK cluster with IBP greater than or equal to 2.8
   def usesTopicId: Boolean =
```

Comment:
I was looking into the uses of this and found that this one in `KafkaApis.handleDeleteTopicsRequest`. 

```scala
        if (!config.usesTopicId && topicIdsFromRequest.contains(topic.topicId)) {
          topic.setErrorCode(Errors.UNSUPPORTED_VERSION.code)
          topic.setErrorMessage("Topic IDs are not supported on the server.")
```

I'm having a little trouble understanding why this check is necessary. If we are not using topic ids, then it seems reasonable if we returned UNKNOWN_TOPIC_ID. 

My concern is the following case. Say I am doing a roll of the cluster to upgrade the IBP to a version which supports topicIds. The controller may get upgraded first and begin creating topics with topicIds before the rest of the cluster has an upgraded IBP. Now if the user tries to delete the topic using the topicId, it could be sent to one of the brokers which does not have the bumped IBP. This would cause an UNSUPPORTED_VERSION error, which would be very surprising.

If you agree this is a problem, we can open a separate issue to address it.

Reply:
There was some discussion on this topic. I think the delete request must be handled by the controller though? Let me know if this is not the case.

Reply:
You are right that DeleteTopics must be handled by the active controller. I think the point still stands with a slightly different scenario. The controller might go from a new version with an updated IBP to an old version during the roll. This could happen because of a session timeout. It is an unlikely case to be sure, but still possible. I'd be inclined to remove the check just to take it out of the picture. It doesn't seem to be buying us anything.

Reply:
Hmmm. In that case would we be going to an old version that doesn't support topic IDs though and this behavior is expected? I think the main reason I wanted to differentiate was so a client would know not to keep trying the request with topic IDs on an old version. But maybe this is not common enough to support this error.

Reply:
This is a good topic to decide on for https://github.com/apache/kafka/pull/9590

Reply:
Ah I totally linked the wrong PR here. I meant my topic ID fetch PR